Change return type of getStringUTF8Length

hzongaro opened the pull request:
The return type of the JVM's `getStringUTF8Length` method has changed from `IDATA` to `UDATA`.  This change adjusts the JIT compiler's uses of that method.  In particular, the return type of `TR_FrontEnd::getStringUTF8Length` and its overriding methods change from `intptr_t` to `int32_t`.  Similarly, the `bufferSize` argument of `TR_FrontEnd::getStringUTF8` becomes `uintptr_t` where it was `int32_t`.

The motivation was that the length of a UTF-8 encoded String could be greater than 2<sup>32</sup> bytes, so the length could overflow `intptr_t` on a 32-bit platform.  All uses of `getStringUTF8Length` in the JIT involve signatures, descriptors, method names and class names, which will never be large enough to exceed the range of the `int32_t` type.  This was something that @jpmpapin was good enough to verify.  Just to be cautious, however, this change includes an assertion test that the computed length is in range for the `int32_t` type.

This change also introduces a method, `getStringUTF8UnabbreviatedLength`, that returns a length value of type `uint64_t` if the JIT compiler ever needs to determine the UTF-8 encoded length of an arbitrary String.  The method is currently unused.

In addition, String Builder Transformer uses the result of `getStringUTF8Length` to estimate the `StringBuilder` buffer size needed to accommodate appending a constant `String` to a `StringBuilder`.  That could overestimate the space required.  This has been changed to use `getStringLength` instead, to use the actual lengths of constant `String` objects.  A test has also been added to detect integer overflow of the capacity estimate, aborting the transformation, as `StringBuilder.<init>`
will throw a `NegativeArraySizeException` if the specified capacity is negative.

This pull request requires a coordinated merge with OMR pull request eclipse-omr/omr#7620

hzongaro commented:
@jdmpapin, thank you for your off-line review of an earlier version of these changes.  May I ask you to review this updated version?

jdmpapin commented on runtime/compiler/env/VMJ9.h:
```diff
@@ -1140,8 +1140,29 @@ class TR_J9VMBase : public TR_FrontEnd
+   /**
+    * \brief Returns the number of UTF-8 encoded bytes needed to represent a Java String object.
+    *        The number of bytes needed to UTF-8 encode the String is representable as
+    *        a \c uintptr_t, in general, but this method returns a length of type \c int32_t.
```
`uintptr_t` might not be enough on 32-bit, so in general we'd need `uint64_t`, which I think is why you made that the return type of `getStringUTF8UnabbreviatedLength()`

hzongaro replied:
Oops!  Thanks for catching that.

hzongaro replied:
Fixed in commit a4974d4242c9205873e6ad6498cb9c724c536cde

jdmpapin commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -8409,7 +8409,7 @@ TR_J9ByteCodeIlGenerator::runFEMacro(TR::SymbolReference *symRef)
          uintptr_t methodHandle;
          uintptr_t methodDescriptorRef;
-         intptr_t methodDescriptorLength;
+         uintptr_t methodDescriptorLength;
```
This change, and the similar changes below and in `handleServerMessage()` seem unnecessary

If we're concerned about possible overflow when adding 1 for the NUL terminator to size a buffer, then `uintptr_t` can avoid that, but on 32-bit `intptr_t` will be the same as `int32_t`, so overflow would still be possible. However, I don't think we're in any danger of passing a string remotely long enough to cause that overflow. It might be better to just adjust the assertion in `getStringUTF8Length()` to make sure that the result is always strictly less than the max, which would guarantee that we can always add 1 without overflow

hzongaro replied:
I'm not sure why I had changed this declaration - I'll revert it to `intptr_t`.

Regarding the change below, from `int` to `intptr_t`, that one was intentional - all of the other declarations of `methodDescriptorLength` in this file declare it as `intptr_t`.  I didn't see a reason why that one should be different, so I changed it for consistency.  Similarly for the declarations as type `size_t` and `int` in `handleServerMessage()`.

hzongaro commented on runtime/compiler/env/VMJ9.cpp:
```diff
@@ -5649,18 +5649,31 @@ TR_J9VMBase::getStringCharacter(uintptr_t objectPointer, int32_t index)
 
 char *
-TR_J9VMBase::getStringUTF8(uintptr_t objectPointer, char *buffer, intptr_t bufferSize)
+TR_J9VMBase::getStringUTF8(uintptr_t objectPointer, char *buffer, int32_t bufferSize)
```
Fixed in commit a4974d4242c9205873e6ad6498cb9c724c536cde

hzongaro replied:
I have also adjusted the signature of `TR_FrontEnd::getStringUTF8` in OMR pull request eclipse-omr/omr#7620.

hzongaro commented on runtime/compiler/net/MessageTypes.hpp:
```diff
@@ -131,6 +131,7 @@ enum MessageType : uint16_t
    VM_getObjectClassInfoFromObjectReferenceLocation,
    VM_stackWalkerMaySkipFrames,
    VM_getStringUTF8Length,
+   VM_getStringUTF8UnabbreviatedLength,
```
Removed in commit a4974d4242c9205873e6ad6498cb9c724c536cde

hzongaro commented on runtime/compiler/optimizer/StringBuilderTransformer.cpp:
```diff
@@ -465,14 +474,23 @@ TR::Node* TR_StringBuilderTransformer::findStringBuilderChainedAppendArguments(T
+   // calculation has overflowed, so halt early.
+   //
+   for (TR_Pair<TR::Node*, TR::RecognizedMethod>* pair = iter.getFirst();
+        (pair != NULL) && (capacity < std::numeric_limits<int32_t>::max());
```
Corrected in commit 1921662c541ad8d3b8f9616013c5ac9907c262a7

hzongaro commented:
Thanks for your review, @jdmpapin.  I have addressed your review comments.  May I ask you to rereview?  Once you are OK with the changes, I will squash the extra commits.

jdmpapin reviewed (approved):
LGTM, please squash

hzongaro commented:
Thanks, @jdmpapin!  I have squashed the commits.

jdmpapin commented:
Jenkins test sanity.functional,sanity.openjdk all jdk8,jdk11,jdk17,jdk21 depends eclipse-omr/omr#7620

hzongaro commented:
It looks like there is a known problem &mdash; issue #21066 &mdash; affecting JITServer tests.  I will rerun PR testing once that problem has been resolved.

jdmpapin commented:
#21066 was closed with a fix this morning

Jenkins test sanity.functional,sanity.openjdk all jdk8,jdk11,jdk17,jdk21 depends eclipse-omr/omr#7620

jdmpapin commented:
Jenkins test sanity.functional,sanity.openjdk aix jdk11 depends eclipse-omr/omr#7620

jdmpapin commented:
Jenkins test sanity.openjdk aix jdk21 depends eclipse-omr/omr#7620

jdmpapin commented:
The `AccessDeniedException` on AIX keeps reoccurring, so I'm going to start ignoring those without restarts

jdmpapin commented:
The PPC64LE Linux JDK11 sanity.openjdk failure is #13756

jdmpapin commented:
This Mac test job failed to fetch origin/pr/21005/merge

Jenkins test sanity.functional xmac jdk21 depends eclipse-omr/omr#7620

jdmpapin commented:
The PPC64LE Linux JDK17 sanity.functional failure is #17474

jdmpapin commented:
Some of the AIX jobs failed with `GitException` instead of `AccessDeniedException`, but that's still infrastructure

The x86-64 Windows JDK11 sanity.functional job also hit `AccessDeniedException`, but first both test lists passed

The AIX JDK17 sanity.openjdk failure is #19962

The x86-32 Windows JDK8 sanity.functional failures are #19678 and #19408

The x86-64 Windows JDK21 sanity.openjdk job (reported here as still pending) had its testList_0 cancelled. It looks like it was still making progress but just running very slowly, which can happen sometimes in CI. There were instances of #21000 and #18463 in testList_2

The x86-64 Windows JDK11 sanity.openjdk job (reported here as still pending) similarly had its testList_1 cancelled. I don't see any real failures

jdmpapin commented:
The tests have finished and all failures are accounted for. @hzongaro, could you please rebase to fix the conflict?

hzongaro commented:
> The tests have finished and all failures are accounted for. @hzongaro, could you please rebase to fix the conflict?

@jdmpapin, thank you for investigating all those failures!  Rebased.

jdmpapin commented:
The diffs are identical except for the update to `CommunicationStream::MINOR_NUMBER`. Doing just one cursory build after the rebase

Jenkins compile xlinux jdk17 depends eclipse-omr/omr#7620

jdmpapin commented:
Jenkins build all

jdmpapin commented:
The build segfaulted in `samplerThreadProc()` :face_exhaling: :disappointed:

I really can't imagine it has anything to do with these changes. It doesn't seem to be a known issue, but it happened twice in a row, so I'm sure it will be one in no time...

BradleyWood commented:
@jdmpapin My local builds have been segfaulting with the same backtrace on the latest OMR/OpenJ9 changes. Not sure exactly what the issue is but its from the last ~20 commits.

jdmpapin commented:
Waiting for OMR to promote

jdmpapin commented:
Repeating the cursory build

Jenkins compile xlinux jdk17 depends https://github.com/eclipse-omr/omr/pull/7620
